### PR TITLE
[Proof Of Concept] POP3 server backed by MessageId and non-LWT based UID/ModSeq generation

### PR DESCRIPTION
**Pull request by chibenwa**
# Why

James POP3 implementation is backed by the IMAP UID, a monotic counter. Cassandra implementation uses LightWeight transactions to back a compare and swap.

This implementation have expensive run time costs, especially in a multi data center setup (LWTs requires 4 round trips accross replica even for reads).

# How

We should contribute to apache/james-project an alternative implementation of the POP3 server not
leveraging UIDs but instead using messageIds, enabling safe, easy to configure multi-datacenter POP3 setup for the Distributed Server.

Use a dedicated view in order to list the messageIds within a mailbox and the size of the messages and use
the messageIdManager to retrieve the given emails.

Offer a configuration option to choose between the classic 'uid based' implementation or the 'messageId based' implementation via the mechanism of module-choosing.

#  Consequences

By implementing this we will have more options in the face of bad Lightweight Transaction performances:

 - Use of `-Dcassandra.unsafe.disable-serial-reads-linearizability=true` option with Cassandra would be acceptable for POP3 workload but it would lead to data loss in IMAP.
 - More aggressivley the tailor made server could include a UID/ModSeq allocation compare and swap mechanism not being backed by LWTs. Trivialy exposed to data races, it 
simulates in the absence of concurrency a correct behaviour without the costs of LightWeight transactions and could timely be setted up. It would lead to data loss in IMAP.

Also POP3 workload would not need a failover script to increment UIDs upon failover. However not playing it timely would result in data loss with IMAP.

All places currently relying on IMAP UID as an identifier will not be able to be relied upon. Impact studies shows that, for some tasks like mail re-indexing will be impacted (as they are backed by IMAP UID),  which is likely non critical for a pure POP3 usage.

# Tests

The following changeset had been tested with Thunderbird (for the Cassandra set up)

### CODE REVIEW

**Comment by chibenwa** [comment]
Can be considered :green_apple:

File: server/protocols/protocols-pop3-distributed/src/test/java/org/apache/james/pop3server/Pop3MetadataStoreContract.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        assertThatCode(() -> Mono.from(testee().clear(generateMailboxId())).block())
+            .doesNotThrowAnyException();
+    }
+
+    @Test
+    default void removeShouldBeIdempotent() {
+        StatMetadata metadata1= new StatMetadata(generateMessageId(), SIZE_1);
+        assertThatCode(() -> Mono.from(testee().remove(generateMailboxId(), metadata1)).block())
+            .doesNotThrowAnyException();
+    }
```
Comment (Arsnael) [inline server/protocols/protocols-pop3-distributed/src/test/java/org/apache/james/pop3server/Pop3MetadataStoreContract.java:132]: Shouldn't add be idempotent too?

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/CassandraRabbitMQJamesServerMain.java [hunk 2]
```diff
@@ -190,6 +190,7 @@ public static GuiceJamesServer createServer(CassandraRabbitMQJamesConfiguration
             .combineWith(BlobStoreCacheModulesChooser.chooseModules(blobStoreConfiguration))
             .combineWith(SearchModuleChooser.chooseModules(searchConfiguration))
             .combineWith(new UsersRepositoryModuleChooser(new CassandraUsersRepositoryModule())
-                .chooseModules(configuration.getUsersRepositoryImplementation()));
+                .chooseModules(configuration.getUsersRepositoryImplementation()))
+            .overrideWith(new DistributedPop3Module());
```
Comment (Arsnael) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/CassandraRabbitMQJamesServerMain.java]: Commit title issue: `Bind the Distributed POP3 server within the Distributed POP3 server. `

You mean instead `Bind the Distributed POP3 server within the Distributed server. `?

**Review by Arsnael — COMMENTED** [review]
I might be wrong, but you seem to put the distributed pop3 server by default into the distributed server. But as you said, that implem could lead to data loss in JMAP if both are used? 

It's not everybody that can easily tweak James to his needs so I would like to be sure on what we get as a default running James here.

Also, I know it's still a POC, but doc would definitely be needed before potentially being merged?

**Comment by chibenwa** [comment]
Yes, this branch bundles the messageId POP3 server within the Distributed server.

This is to ease testing.

If it makes sense performance wise AND If we want to offer that as part of the Apache James server, then that work will need to be reworked to fit in the "module chooser" framework.

Does it answer your question?

**Review by LanKhuat — COMMENTED** [review]
Nothing to add

**Comment by Arsnael** [comment]
> Does it answer your question?

Yes thanks. Nothing to add on the review after the fixups.

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/4321

**Pull request closed without merging**